### PR TITLE
Move selection methods to Writer, protect DocumentSelection

### LINKED ISSUE #1209: Move selection methods to writer
We decided that writer should be the only proper way to modify model and it should include  selection changes. We had plenty of bugs because of the selection changes out of the `enqueueChanges` block, and it should be fixed.

At this point we have following methods to modify document selection:
```js
addRange( range, isBackward = false )
removeAllRanges()
setRanges( newRanges, isLastBackward = false )
setTo( selectable )
setIn( element )
setOn( item )
setCollapsedAt( itemOrPosition, offset )
collapseToStart()
collapseToEnd()
moveFocusTo( itemOrPosition, offset )
removeAttribute( key )
setAttribute( key, value )
setAttributesTo( attrs )
refreshAttributes()
```

We should use this occasion to rethink if we can limit this list. Then selection methods should be changed to protected (and starts from underscore) and proper writter methods should be added.

At the same time the question is if only document selection modification should be protected or any selection.

### PR DESCRIPTION
**Pull request by ma2ciek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

other: Moved selection methods to `Writer`, introduced `LiveSelection`. Closes #1209.

### CODE REVIEW

File: src/model/documentselection.js [hunk 1]
```diff
@@ -7,28 +7,20 @@
...
+import mix from '@ckeditor/ckeditor5-utils/src/mix';
+import EmitterMixin from '@ckeditor/ckeditor5-utils/src/emittermixin';
+import LiveSelection from './liveselection';
 
 /**
  * `DocumentSelection` is a special selection which is used as the
  * {@link module:engine/model/document~Document#selection document's selection}.
  * There can be only one instance of `DocumentSelection` per document.
  *
+ * `DocumentSelection` is a proxy to {@link module:engine/model/liveselection~LiveSelection} that provides
```
Comment (pjasiun) [inline src/model/documentselection.js]: Live Selection should not be public and should not be mentioned in the documentation for public classes.

File: src/model/documentselection.js [hunk 2]
```diff
@@ -44,731 +36,351 @@ const attrOpTypes = new Set(
...
 	 * @param {module:engine/model/document~Document} doc Document which owns this selection.
 	 */
 	constructor( doc ) {
-		super();
-
 		/**
-		 * Document which owns this selection.
+		 * TODO
 		 *
 		 * @protected
```
Comment (pjasiun) [inline src/model/documentselection.js]: Why not private?

File: src/model/documentselection.js [hunk 3]
```diff
@@ -44,731 +36,351 @@ const attrOpTypes = new Set(
...
 	 * Creates an empty live selection for given {@link module:engine/model/document~Document}.
 	 *
 	 * @param {module:engine/model/document~Document} doc Document which owns this selection.
 	 */
 	constructor( doc ) {
-		super();
-
 		/**
-		 * Document which owns this selection.
+		 * TODO
```
Comment (pjasiun) [inline src/model/documentselection.js]: TODO

File: src/model/liveselection.js [hunk 4]
```diff
@@ -0,0 +1,674 @@
+/**
```
Comment (pjasiun) [inline src/model/liveselection.js]: I would make `LiveSelection` private or protected. It should be used only by the `DocumentSelection` and could be even defined there.

File: src/model/utils/deletecontent.js [hunk 5]
```diff
@@ -82,7 +83,11 @@ export default function deleteContent( model, selection, options = {} ) {
 			schema.removeDisallowedAttributes( startPos.parent.getChildren(), writer );
 		}
 
-		selection.setCollapsedAt( startPos );
+		if ( selection instanceof DocumentSelection ) {
+			selection._setTo( startPos );
```
Comment (pjasiun) [inline src/model/utils/deletecontent.js]: The writer should be used here.

File: src/model/utils/deletecontent.js [hunk 6]
```diff
@@ -186,7 +191,12 @@ function insertParagraph( writer, position, selection ) {
 	const paragraph = writer.createElement( 'paragraph' );
 
 	writer.insert( paragraph, position );
-	selection.setCollapsedAt( paragraph );
+
+	if ( selection instanceof DocumentSelection ) {
+		selection._setTo( paragraph );
```
Comment (pjasiun) [inline src/model/utils/deletecontent.js]: As above.

File: src/model/utils/insertcontent.js [hunk 7]
```diff
@@ -54,7 +55,11 @@ export default function insertContent( model, content, selection ) {
 
 		/* istanbul ignore else */
 		if ( newRange ) {
-			selection.setRanges( [ newRange ] );
+			if ( selection instanceof DocumentSelection ) {
+				selection._setTo( newRange );
```
Comment (pjasiun) [inline src/model/utils/insertcontent.js]: As above.

File: src/model/utils/modifyselection.js [hunk 8]
```diff
@@ -64,7 +65,11 @@ export default function modifySelection( model, selection, options = {} ) {
 		const position = tryExtendingTo( data, next.value );
 
 		if ( position ) {
-			selection.moveFocusTo( position );
+			if ( selection instanceof DocumentSelection ) {
+				selection._setFocus( position );
```
Comment (pjasiun) [inline src/model/utils/modifyselection.js]: As above.

File: src/model/writer.js [hunk 9]
```diff
@@ -790,7 +791,139 @@ export default class Writer {
...
+	 * 		// Sets range at the given position.
+	 *		const position = new Position( root, path );
+	 *		selection.setTo( position );
+	 *
+	 * 		// Sets range at the given element.
+	 *		const paragraph = writer.createElement( 'paragraph' );
+	 *		selection.setTo( paragraph, offset );
+	 *
+	 * 		// Removes all ranges.
+	 *		selection.setTo( null );
```
Comment (pjasiun) [inline src/model/writer.js]: `selection.setTo` -> `writer. setSelection`

File: src/model/liveselection.js (hunk 4 above)
Comment (pjasiun) [reply src/model/liveselection.js]: Both classed could go to the same file.

File: src/model/documentselection.js [hunk 10]
```diff
@@ -44,44 +48,406 @@ const attrOpTypes = new Set(
...
 	 * @param {module:engine/model/document~Document} doc Document which owns this selection.
 	 */
 	constructor( doc ) {
-		super();
-
 		/**
-		 * Document which owns this selection.
+		 * Selection used internally by that class (`DocumentSelection` is a proxy to that selection).
 		 *
 		 * @protected
```
Comment (pjasiun) [inline src/model/documentselection.js]: Shouldn't this be private?

File: src/model/documentselection.js [hunk 11]
```diff
@@ -44,44 +48,406 @@ const attrOpTypes = new Set(
...
+	 *
+	 * The location can be specified in the same form as {@link module:engine/model/position~Position.createAt} parameters.
+	 *
+	 * @see module:engine/model/writer~Writer#setSelectionFocus
+	 * @protected
+	 * @param {module:engine/model/item~Item|module:engine/model/position~Position} itemOrPosition
+	 * @param {Number|'end'|'before'|'after'} [offset] Offset or one of the flags. Used only when
+	 * first parameter is a {@link module:engine/model/item~Item model item}.
+	 */
+	_setFocus( itemOrPosition, offset ) {
```
Comment (pjasiun) [inline src/model/documentselection.js]: All protected method should go after all public methods.

File: src/model/documentselection.js [hunk 12]
```diff
@@ -44,44 +48,406 @@ const attrOpTypes = new Set(
...
+	}
+
+	/**
+	 * Unbinds all events previously bound by document selection.
+	 */
+	destroy() {
+		this._selection.destroy();
+	}
+
+	getAttributeKeys() {
```
Comment (pjasiun) [inline src/model/documentselection.js]: Docs?

File: src/model/writer.js [hunk 13]
```diff
@@ -790,8 +791,142 @@ export default class Writer {
...
+	 */
+	setSelectionFocus( itemOrPosition, offset ) {
+		this._assertWriterUsedCorrectly();
+
+		this.model.document.selection._setFocus( itemOrPosition, offset );
+	}
+
+	/**
+	 * Sets attribute(s) on the selection. If attribute with the same key already is set, it's value is overwritten.
+	 *
```
Comment (pjasiun) [inline src/model/writer.js]: Samples needed.

File: src/model/documentselection.js (hunk 10 above)
Comment (ma2ciek) [reply src/model/documentselection.js]: It's a very useful property in tests.

**Comment by pjasiun** [comment]
Selection constructor could use `setTo` and accepts all its parameters. Then we should be able to remove `Selection.createFromSelection`.

File: src/model/writer.js (hunk 13 above)
Comment (ma2ciek) [reply src/model/writer.js]: Fixed.

**Pull request merged**